llarp/link/link_manager.cpp

```diff
@@ -427,6 +432,232 @@ namespace llarp
+
+      {
+        auto rc_bt_list = resp.append_list("rcs");
+
+        const auto& last_time = node_db->get_last_rc_update_times();
+
+        // if since_time isn't epoch start, subtract a bit for buffer
+        if (since_time != decltype(since_time)::min())
+          since_time -= 5s;
+
+        for (const auto& [_, rc] : rcs)
+        {
+          if (last_time.at(rc.router_id()) > since_time or explicit_relays.count(rc.router_id()))
+            rc_bt_list.append_encoded(rc.view());
+        }
+      }
+
+      resp.append("time", now.time_since_epoch().count());
+
+      m.respond(std::move(resp).str(), false);
+    }
+    catch (const std::exception& e)
+    {
+      log::info(link_cat, "Exception handling RC Fetch request: {}", e.what());
+      m.respond(messages::ERROR_RESPONSE);
+    }
+  }
+
+  void
+  LinkManager::fetch_router_ids(const RouterID& source)
+  {
+    if (ep.conns.empty())
+    {
+      log::debug(link_cat, "Not attempting to fetch Router IDs: not connected to any relays.");
+      return;
+    }
+    // TODO: randomize?  Also, keep track of successful responses and drop this edge
+    //       if not many come back successfully.
+    RouterID edge = ep.conns.begin()->first;
+    send_control_message(
+        edge,
+        "fetch_router_ids"s,
+        RouterIDFetch::serialize(source),
+        [this, source = source, edge = std::move(edge)](oxen::quic::message m) {
+          if (not m)
+          {
+            log::info(
+                link_cat,
+                "Error fetching RouterIDs from source \"{}\" via edge \"{}\"",
+                source,
+                edge);
+            node_db->ingest_router_ids(edge, {});  // empty response == failure
+            return;
+          }
+          try
+          {
+            oxenc::bt_dict_consumer btdc{m.body()};
+            btdc.required("routers");
+            auto router_id_strings = btdc.consume_list<std::vector<ustring>>();
+            btdc.require_signature("signature", [&edge](ustring_view msg, ustring_view sig) {
+              if (sig.size() != 64)
+                throw std::runtime_error{"Invalid signature: not 64 bytes"};
+              if (not crypto::verify(edge, msg, sig))
+                throw std::runtime_error{
+                    "Failed to verify signature for fetch RouterIDs response."};
+            });
+            std::vector<RouterID> router_ids;
+            for (const auto& s : router_id_strings)
+            {
+              if (s.size() != RouterID::SIZE)
+              {
+                log::warning(link_cat, "Got bad RouterID from edge \"{}\".", edge);
+                return;
+              }
+              router_ids.emplace_back(s.data());
+            }
+            node_db->ingest_router_ids(edge, std::move(router_ids));
+            return;
+          }
+          catch (const std::exception& e)
+          {
+            log::info(link_cat, "Error handling fetch RouterIDs response: {}", e.what());
```

Comment:
What do you want to do in this error case? Roll on directly into `node_db->ingest_router_ids`? Or re-throw the exception

Reply:
The throw could come from handling bad data (and almost always will barring memory issues, but we can have multiple catch blocks and different exceptions if we need to at some point).  In that case we don't really want the exception to bubble up, but we do still need to signal to node_db somehow that the request to that relay failed.  Using `ingest_router_ids` with an empty result to signal that now, but we can have an explicit failure method if that seems more correct.